swaybar/tray/item.c

```diff
@@ -493,24 +493,36 @@ uint32_t render_sni(cairo_t *cairo, struct swaybar_output *output, double *x,
 		cairo_destroy(cairo_icon);
 	}
 
-	int padded_size = icon_size + 2*padding;
-	*x -= padded_size;
-	int y = floor((height - padded_size) / 2.0);
+	double descaled_padding = (double)padding / output->output_scale;
+	double descaled_icon_size = (double)icon_size / output->output_scale;
+
+	int size = descaled_icon_size + 2 * descaled_padding;
+	*x -= size;
+	int icon_y = floor((output->height - size) / 2.0);
```

Comment:
Note that this calcuates in surface coord, while #6517 calculates in device pixel coord (as before scale simplification). I think this is the source of that vertical alignment difference.

Reply:
I agree. I think the maths in my branch here can be simplified/reduced with some careful thought but that seemed like yak-shaving compared to other things in `swaybar` design that could be improved.

The pixel-specified `tray_padding`, and there-in `output_scale` is necessarily spread through the icon-wrangling and painting works.

The logic is two-fold in that **after finding the best "base" icon for the current height** (an interative process repeating as the bar inits/populates/grows/operates), it is:
1. scaled as close as possible to the desired height avoiding artifacting/aliasing (using only power-of-two enlargement or resampling reduction), while account for and enabling pixel-specified `tray_padding`
1. aligned centered vertically within the (bar) height, again while respecting `tray_padding` (e.g. max-height implications, reaching all the way back to icon generation supporting "hiDPI"/scaling)

I don't know the implications of saving/restoring a cairo state, but I'm betting it's more than the added `cairo_matrix_t` operation.

The only thing I still don't quite wrap my head around is the needing the translation for both the source (surface) pattern and again in the `cairo_rectangle` call. Maybe it makes sense if the scale is 1 but .. I digress; this is why there's probably a (repeated) factor that can be removed in my maths if someone cares.

Reply:
>  the implications of saving/restoring a cairo state

side note: in #6517, they can be replaced by scale(1/scale) and scale(scale) back, but that involves float points, so I save the states and reset transformation matrix to identical matrix to avoid that. (An implication is that additional transformation, like translation, will also be ignored during icon rendering, but there is none and the current design passes `x` all the way thus translation is likely not needed)

My concern is that aligning icon+padding to surface coord may look bad on fractional scale, creating unneeded aliasing. Aligning to device pixels should be better (and possibly more accurate on making it center).

Reply:
As well, I don't think the code-paths properly support fractional scaling; at least not in swaybar.